File: packages/orchestrator/src/actors/agent/tool-executor.ts
// ---- AgentActor — tool execution (parallel / sequential) ----

import {
  type DebugTraceInput,
  debugTrace,
  type Message,
  type ModelRunSettings,
  startDebugSpan,
  type ToolExecResult,
  type ToolExecutionContext,
} from "piko-orchestrator-protocol";
import type { ActorContext } from "../../kernel/actor-system.js";
import type { CatalogRoute } from "../../tools/tool-registry.js";
import type { AgentActorDeps, AgentRuntimeState, AgentWorkerState } from "./types.js";

function createAbortError(): Error {
  const error = new Error("Task cancelled");
  error.name = "AbortError";
  return error;
}

/**
 * Stop awaiting an operation as soon as cancellation is requested, even when
 * the underlying provider ignores the AbortSignal. The provider promise is
 * still observed so a late rejection cannot become unhandled.
 */
export function raceWithAbort<T>(
  operation: Promise<T>,
  signal?: AbortSignal,
  onLateSettlement?: (outcome: "completed" | "error") => void,
): Promise<T> {
  if (!signal) return operation;
  if (signal.aborted) return Promise.reject(createAbortError());

  return new Promise<T>((resolve, reject) => {
    let settled = false;

    const finish = (callback: () => void) => {
      if (settled) return;
      settled = true;
      signal.removeEventListener("abort", onAbort);
      callback();
    };
    const onAbort = () => finish(() => reject(createAbortError()));

    signal.addEventListener("abort", onAbort, { once: true });
    // Cover cancellation between the initial check and listener registration.
    if (signal.aborted) onAbort();

    operation.then(
      (value) => {
        if (settled) onLateSettlement?.("completed");
        finish(() => resolve(value));
      },
      (error) => {
        if (settled) onLateSettlement?.("error");
        finish(() => reject(error));
      },
    );
  });
}

async function executeOneTool(
  deps: AgentActorDeps,
  tc: { id: string; name: string; arguments: Record<string, unknown> },
  execContext: ToolExecutionContext,
  route: CatalogRoute,
  signal?: AbortSignal,
): Promise<ToolExecResult> {
  const fields: Partial<DebugTraceInput> = {
    taskId: execContext.taskId,
    agentId: execContext.agentId,
    toolCallId: tc.id,
    toolName: tc.name,
    eventSeq: execContext.eventSeq,
    signalAborted: signal?.aborted ?? false,
  };
  const span = startDebugSpan("tool.execute", fields);
  try {
    const result = await raceWithAbort(
      deps.toolRegistry.executeTool(
        { type: "toolCall", id: tc.id, name: tc.name, arguments: tc.arguments },
        execContext,
        route,
        signal,
      ),
      signal,
      (outcome) => debugTrace({ ...fields, stage: "tool.execute.late_settlement", outcome }),
    );
    span.end({ outcome: result.error ? "error" : "completed" });
    return result;
  } catch (error) {
    span.end({
      outcome:
        signal?.aborted || (error instanceof Error && error.name === "AbortError")
          ? "aborted"
          : "error",
      signalAborted: signal?.aborted ?? false,
    });
    throw error;
  }
}

// ---- Execution mode resolution ----

/**
 * Determine whether the batch of tool calls should run in parallel or sequentially.
 *
 * Rules:
 * 1. If settings.parallelTools === false → sequential (config override)
 * 2. If any tool in the batch has executionMode === "sequential" → sequential
 * 3. Otherwise → parallel (default)
 */
export function resolveExecutionMode(
  toolCalls: Array<{ name: string }>,
  routes: Map<string, CatalogRoute>,
  settings: ModelRunSettings,
): "parallel" | "sequential" {
  if (settings.parallelTools === false) return "sequential";

  for (const tc of toolCalls) {
    const route = routes.get(tc.name);
    if (route?.toolDef?.executionMode === "sequential") return "sequential";
  }

  return "parallel";
}

// ---- Tool execution ----

/**
 * Execute a batch of tool calls.
 * Parallel path: executes all tool calls concurrently via Promise.all.
 * Sequential path: executes each tool call sequentially in a for...of loop.
 *
 * Results are always appended to transcript in the original tool call order.
 * Ordering metadata (turnIndex, parentMessageId, contentIndex, toolCallIndex)
 * is passed through the ToolExecutionContext for the tool-registry to emit.
 */
export async function executeToolCalls(
  state: AgentRuntimeState,
  workerState: AgentWorkerState,
  deps: AgentActorDeps,
  ctx: ActorContext,
  taskId: string,
  toolCalls: Array<{
    id: string;
    name: string;
    arguments: Record<string, unknown>;
  }>,
  modelSettings: ModelRunSettings,
  routes: Map<string, CatalogRoute>,
  signal?: AbortSignal,
  parentMessageId?: string,
  toolCallOrder?: Map<string, { contentIndex: number; toolCallIndex: number }>,
): Promise<void> {
  const mode = resolveExecutionMode(toolCalls, routes, modelSettings);

  if (mode === "parallel") {
    return executeParallel(
      state,
      workerState,
      deps,
      ctx,
      taskId,
      toolCalls,
      routes,
      signal,
      parentMessageId,
      toolCallOrder,
    );
  }

  return executeSequential(
    state,
    workerState,
    deps,
    ctx,
    taskId,
    toolCalls,
    routes,
    signal,
    parentMessageId,
    toolCallOrder,
  );
}

/** Build execution context with ordering metadata. */
function makeContext(
  state: AgentRuntimeState,
  taskId: string,
  tc: { id: string },
  turnIndex: number,
  eventSeq: number,
  nextEventSeq: () => number,
  parentMessageId?: string,
  toolCallOrder?: Map<string, { contentIndex: number; toolCallIndex: number }>,
): ToolExecutionContext {
  const order = toolCallOrder?.get(tc.id);
  return {
    agentId: state.spec.id,
    taskId,
    toolSetIds: state.spec.toolSetIds,
    turnIndex,
    eventSeq,
    nextEventSeq,
    parentMessageId: parentMessageId ?? "",
    contentIndex: order?.contentIndex ?? 0,
    toolCallIndex: order?.toolCallIndex ?? 0,
  };
}

/** Parallel execution: runs all tool calls concurrently. */
async function executeParallel(
  state: AgentRuntimeState,
  workerState: AgentWorkerState,
  deps: AgentActorDeps,
  _ctx: ActorContext,
  taskId: string,
  toolCalls: Array<{
    id: string;
    name: string;
    arguments: Record<string, unknown>;
  }>,
  routes: Map<string, CatalogRoute>,
  signal?: AbortSignal,
  parentMessageId?: string,
  toolCallOrder?: Map<string, { contentIndex: number; toolCallIndex: number }>,
): Promise<void> {
  // Fire all execution requests concurrently
  const promises = toolCalls.map(async (tc, index) => {
    const route = routes.get(tc.name);
    if (!route) {
      return { index, tc, error: `No route for tool "${tc.name}"` };
    }

    if (signal?.aborted) {
      return { index, tc, error: "Task cancelled" };
    }

    const execContext = makeContext(
      state,
      taskId,
      tc,
      workerState.stepCount,
      workerState.eventSeq,
      () => ++workerState.eventSeq,
      parentMessageId,
      toolCallOrder,
    );

    try {
      const execResult = await executeOneTool(deps, tc, execContext, route, signal);
      return { index, tc, result: execResult };
    } catch (err) {
      return {
        index,
        tc,
        result: { error: err instanceof Error ? err.message : String(err) },
      };
    }
  });

  const results = await Promise.all(promises);

  // Append results in original order
  results.sort((a, b) => a.index - b.index);
  for (const item of results) {
    if ("error" in item) {
      workerState.transcript.push({
        role: "toolResult",
        toolName: item.tc.name,
        toolCallId: item.tc.id,
        content: [{ type: "text", text: `Tool error: ${item.error}` }],
        details: { error: item.error },
        isError: true,
        timestamp: Date.now(),
      } as Message);
    } else {
      appendToolResult(workerState, item.tc, item.result as ToolExecResult);
    }
  }
}

async function executeSequential(
  state: AgentRuntimeState,
  workerState: AgentWorkerState,
  deps: AgentActorDeps,
  _ctx: ActorContext,
  taskId: string,
  toolCalls: Array<{
    id: string;
    name: string;
    arguments: Record<string, unknown>;
  }>,
  routes: Map<string, CatalogRoute>,
  signal?: AbortSignal,
  parentMessageId?: string,
  toolCallOrder?: Map<string, { contentIndex: number; toolCallIndex: number }>,
): Promise<void> {
  for (const tc of toolCalls) {
    const execContext = makeContext(
      state,
      taskId,
      tc,
      workerState.stepCount,
      workerState.eventSeq,
      () => ++workerState.eventSeq,
      parentMessageId,
      toolCallOrder,
    );

    if (signal?.aborted) {
      workerState.transcript.push({
        role: "toolResult",
        toolName: tc.name,
        toolCallId: tc.id,
        content: [{ type: "text", text: "Tool error: Task cancelled" }],
        details: { error: "Task cancelled" },
        isError: true,
        timestamp: Date.now(),
      } as Message);
      continue;
    }

    const route = routes.get(tc.name);
    if (!route) {
      workerState.transcript.push({
        role: "toolResult",
        toolName: tc.name,
        toolCallId: tc.id,
        content: [{ type: "text", text: `Tool error: No route for tool "${tc.name}"` }],
        details: { error: `No route for tool "${tc.name}"` },
        isError: true,
        timestamp: Date.now(),
      } as Message);
      continue;
    }

    try {
      const execResult = await executeOneTool(deps, tc, execContext, route, signal);
      appendToolResult(workerState, tc, execResult);
    } catch (err) {
      const errorText = err instanceof Error ? err.message : String(err);
      workerState.transcript.push({
        role: "toolResult",
        toolName: tc.name,
        toolCallId: tc.id,
        content: [{ type: "text", text: `Tool error: ${errorText}` }],
        details: { error: errorText },
        isError: true,
        timestamp: Date.now(),
      } as Message);
    }
  }
}

/** Append a successful tool execution result to the transcript. */
export function appendToolResult(
  workerState: AgentWorkerState,
  tc: { id: string; name: string },
  execResult: ToolExecResult,
): void {
  const text =
    typeof execResult.value === "string"
      ? execResult.value
      : JSON.stringify(execResult.ok ? execResult.value : execResult.error, null, 2);

  workerState.transcript.push({
    role: "toolResult",
    toolName: tc.name,
    toolCallId: tc.id,
    content: [{ type: "text", text }],
    details: execResult.ok ? execResult.value : execResult.error,
    isError: !execResult.ok,
    timestamp: Date.now(),
  } as Message);
}

File: packages/orchd/src/actors/agent/tool_executor.rs
// ---- AgentActor: tool executor — parallel / sequential execution ----

use std::collections::HashMap;
use std::sync::Arc;

use tokio::sync::Mutex;
use tokio_util::sync::CancellationToken;

use crate::protocol::messages::{ContentBlock, Message};
use crate::protocol::model::ModelRunSettings;
use crate::protocol::tools::ToolExecutionMode;
use crate::tools::registry::{CatalogRoute, ToolRegistry};
use crate::tools::{ToolExecError, ToolExecResult, ToolExecutionContext};

use super::types::*;

/// Generate a stable runtime tool entity ID.
pub(crate) fn runtime_tool_entity_id(parent_message_id: &str, tool_call_index: u32) -> String {
    format!("{}:tool:{}", parent_message_id, tool_call_index)
}


// ---- Tool call descriptor ----

#[derive(Debug, Clone)]
pub struct ToolCallItem {
    pub content_index: u32,
    pub tool_call_index: u32,
    pub id: String,
    pub name: String,
    pub arguments: serde_json::Value,
}

// ---- Public API ----

/// Execute a batch of tool calls. Resolves parallel vs sequential automatically.
#[tracing::instrument(skip(state, worker, deps, tool_calls, routes, cancel), fields(task_id = %task_id, tool_count = tool_calls.len()))]
pub async fn execute_tool_calls(
    state: &Arc<Mutex<AgentRuntimeState>>,
    worker: &mut AgentWorkerState,
    deps: &AgentActorDeps,
    task_id: &str,
    tool_calls: &[ToolCallItem],
    model_settings: &ModelRunSettings,
    routes: &HashMap<String, CatalogRoute>,
    cancel: CancellationToken,
    parent_message_id: &str,
) {
    let mode = resolve_execution_mode(tool_calls, routes, model_settings);

    if mode == ExecutionMode::Parallel {
        execute_parallel(
            state,
            worker,
            deps,
            task_id,
            tool_calls,
            routes,
            cancel,
            parent_message_id,
        )
        .await;
    } else {
        execute_sequential(
            state,
            worker,
            deps,
            task_id,
            tool_calls,
            routes,
            cancel,
            parent_message_id,
        )
        .await;
    }
}

#[derive(PartialEq)]
enum ExecutionMode {
    Parallel,
    Sequential,
}

fn resolve_execution_mode(
    calls: &[ToolCallItem],
    routes: &HashMap<String, CatalogRoute>,
    settings: &ModelRunSettings,
) -> ExecutionMode {
    if settings.parallel_tools == Some(false) {
        return ExecutionMode::Sequential;
    }

    for tc in calls {
        if let Some(route) = routes.get(&tc.name)
            && matches!(
                &route.tool_def.execution_mode,
                Some(ToolExecutionMode::Sequential)
            )
        {
            return ExecutionMode::Sequential;
        }
    }

    ExecutionMode::Parallel
}

async fn get_agent_id(state: &Arc<Mutex<AgentRuntimeState>>) -> String {
    state.lock().await.spec.id.clone()
}

// ---- Parallel execution ----

async fn execute_parallel(
    state: &Arc<Mutex<AgentRuntimeState>>,
    worker: &mut AgentWorkerState,
    deps: &AgentActorDeps,
    task_id: &str,
    tool_calls: &[ToolCallItem],
    routes: &HashMap<String, CatalogRoute>,
    cancel: CancellationToken,
    parent_message_id: &str,
) {
    let agent_id = get_agent_id(state).await;
    let host_context = state.lock().await.current_host_context.clone();
    let turn_index = worker.step_count;

    let mut handles = Vec::new();

    for tc in tool_calls.iter() {
        let route = routes.get(&tc.name).cloned();
        let deps = deps.clone();
        let agent_id = agent_id.clone();
        let host_context = host_context.clone();
        let task_id = task_id.to_string();
        let tc = tc.clone();
        let pm_id = parent_message_id.to_string();
        let cancel = cancel.clone();

        let handle = tokio::spawn(async move {
            if cancel.is_cancelled() {
                return (
                    tc.clone(),
                    ToolExecResult {
                        ok: false,
                        value: None,
                        error: Some(ToolExecError {
                            code: "cancelled".into(),
                            message: "Task cancelled".into(),
                            retryable: Some(false),
                        }),
                    },
                );
            }

            match route {
                Some(r) => {
                    let call = ContentBlock::ToolCall {
                        id: tc.id.clone(),
                        name: tc.name.clone(),
                        arguments: tc.arguments.clone(),
                        partial_json: None,
                    };
                    let exec_ctx = ToolExecutionContext {
                        agent_id: agent_id.clone(),
                        task_id: task_id.clone(),
                        tool_set_ids: vec![],
                        turn_index: Some(turn_index),
                        event_seq: Some(0),
                        next_event_seq: None,
                        parent_message_id: Some(pm_id.clone()),
                        content_index: Some(tc.content_index),
                        tool_call_index: Some(tc.tool_call_index),
                        tool_entity_id: Some(runtime_tool_entity_id(&pm_id, tc.tool_call_index)),
                        host_context,
                    };

                    let result = (*deps.tool_registry)
                        .execute_tool(&call, &exec_ctx, &r, Some(cancel.clone()))
                        .await;

                    // execute_tool returns ToolExecResult directly
                    (tc.clone(), result)
                }
                None => {
                    let error_result = ToolExecResult {
                        ok: false,
                        value: None,
                        error: Some(ToolExecError {
                            code: "not_found".into(),
                            message: format!("No route for tool \"{}\"", tc.name),
                            retryable: Some(false),
                        }),
                    };
                    (tc.clone(), error_result)
                }
            }
        });

        handles.push(handle);
    }

    let mut results: Vec<(ToolCallItem, ToolExecResult)> = Vec::new();
    for h in handles {
        if let Ok((tc, result)) = h.await {
            results.push((tc, result));
        }
    }

    results.sort_by_key(|(tc, _)| tc.tool_call_index);

    for (tc, exec_result) in results {
        append_tool_result(worker, &tc, &exec_result);
    }
}

// ---- Sequential execution ----

async fn execute_sequential(
    state: &Arc<Mutex<AgentRuntimeState>>,
    worker: &mut AgentWorkerState,
    deps: &AgentActorDeps,
    task_id: &str,
    tool_calls: &[ToolCallItem],
    routes: &HashMap<String, CatalogRoute>,
    cancel: CancellationToken,
    parent_message_id: &str,
) {
    let agent_id = get_agent_id(state).await;
    let host_context = state.lock().await.current_host_context.clone();
    let turn_index = worker.step_count;

    for tc in tool_calls {
        if cancel.is_cancelled() {
            append_tool_error(worker, tc, "Task cancelled");
            continue;
        }

        let route = match routes.get(&tc.name) {
            Some(r) => r,
            None => {
                append_tool_error(worker, tc, &format!("No route for tool \"{}\"", tc.name));
                continue;
            }
        };

        let call = ContentBlock::ToolCall {
            id: tc.id.clone(),
            name: tc.name.clone(),
            arguments: tc.arguments.clone(),
            partial_json: None,
        };

        let exec_ctx = ToolExecutionContext {
            agent_id: agent_id.clone(),
            task_id: task_id.to_string(),
            tool_set_ids: vec![],
            turn_index: Some(turn_index),
            event_seq: Some(0),
            next_event_seq: None,
            parent_message_id: Some(parent_message_id.to_string()),
            content_index: Some(tc.content_index),
            tool_call_index: Some(tc.tool_call_index),
            tool_entity_id: Some(runtime_tool_entity_id(
                parent_message_id,
                tc.tool_call_index,
            )),
            host_context: host_context.clone(),
        };

        let result = (*deps.tool_registry)
            .execute_tool(&call, &exec_ctx, route, Some(cancel.clone()))
            .await;

        append_tool_result(worker, tc, &result);
    }
}

// ---- Append helpers ----

fn append_tool_result(worker: &mut AgentWorkerState, tc: &ToolCallItem, result: &ToolExecResult) {
    if result.ok {
        let text = match &result.value {
            Some(v) if v.is_string() => v.as_str().unwrap_or("").to_string(),
            Some(v) => serde_json::to_string_pretty(v).unwrap_or_default(),
            None => String::new(),
        };

        worker.transcript.push(Message::ToolResult {
            tool_call_id: tc.id.clone(),
            tool_name: Some(tc.name.clone()),
            content: vec![ContentBlock::Text { text }],
            details: result.value.clone(),
            is_error: Some(false),
            timestamp: None,
        });
    } else {
        let error_msg = result
            .error
            .as_ref()
            .map(|e| e.message.clone())
            .unwrap_or_else(|| "Unknown error".into());

        worker.transcript.push(Message::ToolResult {
            tool_call_id: tc.id.clone(),
            tool_name: Some(tc.name.clone()),
            content: vec![ContentBlock::Text { text: error_msg }],
            details: result
                .error
                .as_ref()
                .map(|e| serde_json::to_value(e).unwrap_or_default()),
            is_error: Some(true),
            timestamp: None,
        });
    }
}

fn append_tool_error(worker: &mut AgentWorkerState, tc: &ToolCallItem, error: &str) {
    worker.transcript.push(Message::ToolResult {
        tool_call_id: tc.id.clone(),
        tool_name: Some(tc.name.clone()),
        content: vec![ContentBlock::Text {
            text: format!("Tool error: {error}"),
        }],
        details: Some(serde_json::json!({"error": error})),
        is_error: Some(true),
        timestamp: None,
    });
}
